experimental/alp/lib/Transforms/extract_kernel_pass.cpp

```diff
@@ -183,4 +194,149 @@ struct ExtractKernelPass : public ExtractKernelPassBase<ExtractKernelPass> {
 
 std::unique_ptr<mlir::OperationPass<ModuleOp>> mlir::createExtractKernelPass() {
   return std::make_unique<ExtractKernelPass>();
-}
\ No newline at end of file
+}
+
+void extract_function_2(StringRef func_name, Block *block, ModuleOp parentModule, RewriterBase &rewriter, Block* origin, Location loc)
+{
+  SmallVector<Type, 1> Input, Output;
+
+  // Create the function (callee site) with an empty block
+  rewriter.setInsertionPointToStart(parentModule.getBody());
+  auto func_op = rewriter.create<FuncOp>(parentModule.getLoc(), func_name, FunctionType::get(parentModule.getContext(), Input, Output));
+  auto entry_block = func_op.addEntryBlock();
+
+  // Build the dominance tree of the parent op of the block
+  Region *region = block->getParent();
+  Operation *parent_op = region->getParentOp();
+  auto dom_info = mlir::DominanceInfo(parent_op);
+
+  //std::set<int64_t> consts;
+  llvm::SmallVector<Value, 1> vals;
+  llvm::SmallVector<arith::ConstantOp, 1> consts;
+
+  // Walk the block and find out all the variables that were defined outside 
+  // this block and are used inside the block (i.e., all the variables x that
+  // properly dominate the block). The only things we will redefine inside the
+  // entry block are constants. For all other variables, we will add them as 
+  // inputs to the function
+  block->walk([&](Operation *inst) {
```

Comment:
I've seen this implemented in various places as:
```
bool isDefinedOutsideOfX(Value value) {
  return !region().isAncestor(value.getParentRegion());
}
```

Reply:
Thanks Nicolas! Can I ask more info? What does this replace? I still need to walk every operation of the block, right?